### DISCUSSION TITLE
[dashboard filter component] Order and Group filters in dashboard

### DISCUSSION THREAD

**Discussion by graceguo-supercat** [apache/superset#16152] [Ideas]
In airbnb we have many advance Superset dashboard builders. They built large dashboard with many charts and filters. Because these large dashboard will be shared between many teams and organization, our dashboard builders also make Markdown component to describe how to use filter(s), here is a good example:
![Screen_Shot_2021-08-05_at_2_34_20_PM](https://user-images.githubusercontent.com/27990562/128761900-2139e5a0-7f53-486c-a57c-fa78c39c05b9.jpg)

But if we enable dashboard native filter components, and migrate to native filters (and remove filter_boxes), dashboard will display all filters in a list on the filters bar without customized order and group, our dashboards usability will be greatly impacted. This is biggest reason that block airbnb move forward to native filters solution.

So here we propose a solution (and in airbnb we started to implement it):
![Screen_Shot_2021-08-05_at_4_17_52_PM-2](https://user-images.githubusercontent.com/27990562/128763503-307d22af-9fb9-4e94-84d5-41c0f61e68fa.jpg)
- For each filter component, we will allow user to add text description. We will show filter description in hover-tooltip or click-popover.

- Besides filter component, we create a `Section component` with title, and body. ~~It will display a title and Markdown content.~~
Body is some simple text.

- Section component doesn't belong to any filter(s). ~~It could be created or used like regular markdown component in dashboard.~~ It is a _divider_ that split a list of filter components into different sections.

- Add function to allow user add `Section component` into filter bar, just like we add filter component.

- When dashboard is in edit mode, allow user to drag-n-drop filter component and section component to arrange the order for display.

- The order of filter and section components will be saved into dashboard metadata.

- When dashboard viewer switches tab, some filter components may be out-of-scope. In this case, Superset has logic to hide out-of-scope filter components, and may have new filter components showing up.  In this case, section component will always stay no change, but filter components between sections could be removed.

I updated proposal based on our meeting. thanks for discussion

cc @junlincc @michael-s-molina @villebro @m-ajay @evans @kgabryje

**Comment by junlincc** [comment]
I think this is useful and would like to work/collaborate on it, maybe not in the form of a markdown, but title description(?). 🤔.  we will assess the difficulty and propose some design/idea. 🙏

**Comment by villebro** [comment]
Some random thoughts:
- Since we now have in scope/out of scope sections on the filter tab, there is probably less need for grouping and instructions if the dashboard layout is intuitively designed. However I do understand the need to be able to add additional context to the filters and potentially group them somehow along with instructions to further improve understandability in complex dashboards.
- Since the markdown component doesn't have a dataset, and therefore doesn't affect any charts, we will need to introduce a new mechanism for letting the filter tab know when the component should be visible/hidden. Perhaps we could introduce a select component in which the user can select which markdown component a filter is tied to, and then make the component visible if at least one filter is visible.
- Like Junlin said, maybe we it would be possible to introduce something alternative, like titles or filter sections with a tooltip that's customizable. This would introduce less clutter on the tab while making it possible to add arbitrarily long instructions when necessary. However, if we introduce sections, the order of filters will also change (all filters will need to be primarily be sorted by section, and only secondarily by the order in the metadata array).

**Comment by michael-s-molina** [comment]
One possible solution to the problem is to allow the user to create filter groups that coexist with the outer of scope filters.

Similar to Power BI fixed groups but with dynamic behavior.

<img width="247" alt="Screen Shot 2021-08-11 at 9 38 53 AM" src="https://user-images.githubusercontent.com/70410625/129041723-b0efb471-61e2-4731-b982-ac88cb21aa63.png">

Taking the example of the discussion a user could create a filter bar with the following groups.
Step 1: Time Range ℹ️ 
Step 2: New Route ℹ️ 
Step 3: Comparison Group ℹ️  
Out of scope (appears when necessary) ℹ️  

Each group could have an associated description that supports markdown and gets exhibited in a tooltip.

Then here we could have an "+Add filter group" button that leads to a screen with the group name and markdown. It would be possible to drag and drop filters to filter groups and sort filters and filter groups.

<img width="894" alt="Screen Shot 2021-08-11 at 10 06 53 AM" src="https://user-images.githubusercontent.com/70410625/129042283-eb816371-0aac-473c-a410-1f4f205dce2e.png">

Questions:

- Can we have a good design for the case where we have a hierarchical filter inside a group? Check [this comment](https://github.com/apache/superset/pull/15433#issuecomment-885685647) for some context on hierarchical filters.

- A filter can also have a description or just the filter group?

**Reply by graceguo-supercat** [reply]
Hi @michael-s-molina,
thanks for your suggestion. `+Add filter group` and group filter components in a **_hierarchical_** structure, instead of current **_list_** structure, was discussed in airbnb (and Preset). We feel it might introduce a lot more complex in implementation and maintain. For the first version that introducing group and order for filter components, we will keep it simple. I think in the meeting we also talked about in the future, Superset will allow user add filter component **_next to charts_** as filter_box did. So this solution, dedicated to solve problems in filter bar, may not last very long.

Based on our discussion in the last meeting, we will allow description for each filter, as well as section. I will update this in the proposal.
